fix(parse): arrange code following regression with tika 3.2.2

fdelbrayelle opened the pull request:
fixes https://github.com/kestra-io/kestra/issues/11362

<!-- Thanks for submitting a Pull Request to kestra. To help us review your contribution, please follow the guidelines below:

- Make sure that your commits follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification e.g. `feat(ui): add a new navigation menu item` or `fix(core): fix a bug in the core model` or `docs: update the README.md`. This will help us automatically generate the changelog.
- The title should briefly summarize the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share a flow example to help the reviewer understand and QA the change.
- Use "close" to automatically close an issue. For example, `close #1234` will close issue #1234. -->

### What changes are being made and why?

After upgrading to Apache Tika 3.2.2, simple PDFs started returning empty `content` when parsed with `AutoDetectParser` and a `PDFParserConfig` set in the `ParseContext`.

This behavior is **not mentioned in the official Tika changelogs** for [3.2.2](https://archive.apache.org/dist/tika/3.2.2/CHANGES-3.2.2.txt) or [3.2.3](https://downloads.apache.org/tika/3.2.3/CHANGES-3.2.3.txt), which mainly reference unrelated detection and POIFS fixes (e.g., TIKA-4441, TIKA-4424).

Empirically, the regression appears between 3.2.1 and 3.2.2 and seems tied to changes in internal stream/detection handling. Using `PDFParser` explicitly for PDF files restores normal text extraction without affecting other formats.

<!-- Please include a brief summary of the changes included in this PR e.g. closes #1234. -->

---

### How the changes have been QAed?

Input PDF:

[Tika_Plugin_Input.pdf](https://github.com/user-attachments/files/24034993/Tika_Plugin_Input.pdf)

Flow:

```yaml
id: tika_pdf_import
namespace: company.team
inputs:
  - id: file
    type: FILE
  - id: contenttype
    type: SELECT
    defaults: TEXT
    values:
      - TEXT
      - XHTML
      - XHTML_NO_HEADER
    isRadio: true
tasks:
  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: "{{ inputs.file }}"
    store: false
    ocrOptions:
      strategy: NO_OCR
    extractEmbedded: false
    contentType: "{{inputs.contenttype}}"
    logLevel: DEBUG
```

Result:

<img width="3042" height="1266" alt="Capture d’écran du 2025-10-13 16-01-59" src="https://github.com/user-attachments/assets/65c290aa-9334-4767-843b-0436cd4cd07f" />

<!-- Include example code that shows how this PR has been QAed. The code should present a complete yet easily reproducible flow.

```yaml
# Your example flow code here
```

Note that this is not a replacement for unit tests but rather a way to demonstrate how the changes work in a real-life scenario, as the end-user would experience them.

Remove this section if this change applies to all flows or to the documentation only. -->

---

### Setup Instructions

<!--If there are any setup requirements like API keys or trial accounts, kindly include brief bullet-points-description outlining the setup process below.

- [External System Documentation](URL)
- Steps to set up the necessary resources

If there are no setup requirements, you can remove this section.

Thank you for your contribution. ❤️  -->

---

### Contributor Checklist ✅

- [ ] PR Title and commits follows [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Add a `closes #ISSUE_ID` or `fixes #ISSUE_ID` in the description if the PR relates to an opened issue.
- [ ] Documentation updated (plugin docs from `@Schema` for properties and outputs, `@Plugin` with examples, `README.md` file with basic knowledge and specifics).
- [ ] Setup instructions included if needed (API keys, accounts, etc.).
- [ ] Prefix all rendered properties by `r` not `rendered` (eg: `rHost`).
- [ ] Use `runContext.logger()` to log enough important infos where it's needed and with the best level (DEBUG, INFO, WARN or ERROR).

⚙️ **Properties**
- [ ] Properties are declared with `Property<T>` carrier type, do **not** use `@PluginProperty`.
- [ ] Mandatory properties must be annotated with `@NotNull` and checked during the rendering.
- [ ] You can model a JSON thanks to a simple `Property<Map<String, Object>>`.

🌐 **HTTP**
- [ ] Must use Kestra’s internal HTTP client from `io.kestra.core.http.client`

📦 **JSON**
- [ ] If you are serializing response from an external API, you may have to add a `@JsonIgnoreProperties(ignoreUnknown = true)` at the mapped class level. So that we will avoid to crash the plugin if the provider add a new field suddenly.
- [ ] Must use Jackson mappers provided by core (`io.kestra.core.serializers`)

✨ **New plugins / subplugins**
- [ ] Make sure your new plugin is configured like mentioned [here](https://kestra.io/docs/plugin-developer-guide/gradle#mandatory-configuration).
- [ ] Add a `package-info.java` under each sub package respecting [this format](https://github.com/kestra-io/plugin-odoo/blob/main/src/main/java/io/kestra/plugin/odoo/package-info.java) and choosing the right category.
- [ ] Icons added in `src/main/resources/icons` in SVG format and not in thumbnail (keep it big):
  - `plugin-icon.svg`
  - One icon per package, e.g. `io.kestra.plugin.aws.svg`
  - For subpackages, e.g. `io.kestra.plugin.aws.s3`, add `io.kestra.plugin.aws.s3.svg`
    See example [here](https://github.com/kestra-io/plugin-elasticsearch/blob/master/src/main/java/io/kestra/plugin/elasticsearch/Search.java#L76).
- [ ] Use `"{{ secret('YOUR_SECRET') }}"` in the examples for sensible infos such as an API KEY.
- [ ] If you are fetching data (one, many or too many), you must add a `Property<FetchType> fetchType` to be able to use `FETCH_ONE`, `FETCH` and even `STORE` to store big amount of data in the internal storage.
- [ ] Align the `"""` to close examples blocks with the flow id.

🧪 **Tests**
- [ ] Unit Tests added or updated to cover the change (using the `RunContext` to actually run tasks).
- [ ] Add sanity checks if possible with a YAML flow inside `src/test/resources/flows`.
- [ ] Avoid disabling tests for CI. Instead, configure a local environment whenever it's possible with `.github/setup-unit.sh` (which can be executed locally and in the CI) all along with a new `docker-compose-ci.yml` file (do **not** edit the existing `docker-compose.yml`).
- [ ] Provide screenshots from your QA / tests locally in the PR description. The goal here is to use the JAR of the plugin and directly test it locally in Kestra UI to ensure it integrates well.

📤 **Outputs**
- [ ] Do not send back as outputs the same infos you already have in your properties.
- [ ] If you do not have any output use `VoidOutput`.
- [ ] Do not output twice the same infos (eg: a status code, an error code saying the same thing...).

Malaydewangan09 commented:
Seems to work fine with PDFs but not with images and text files:

```yaml
id: parse-image-metadata-using-apache-tika
namespace: company.team

tasks:
  - id: get_image
    type: io.kestra.plugin.core.http.Download
    uri: https://kestra.io/blogs/2023-05-31-beginner-guide-kestra.jpg

  - id: tika
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.get_image.uri }}"
    store: false
    contentType: TEXT
    ocrOptions:
      strategy: OCR_AND_TEXT_EXTRACTION
```

```yaml
id: tika_parse
namespace: company.team

inputs:
  - id: file
    type: FILE

tasks:
  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: '{{ inputs.file }}'
    extractEmbedded: true
    store: false
```

 Outputs:
```json
{
  "embedded": {},
  "metadata": {
    "X-TIKA:Parsed-By": "org.apache.tika.parser.EmptyParser",
    "X-TIKA:Parsed-By-Full-Set": "org.apache.tika.parser.EmptyParser",
    "Content-Type": "image/jpeg"
  },
  "content": ""
}
```

<img width="1389" height="557" alt="Screenshot 2025-12-09 at 1 07 54 PM" src="https://github.com/user-attachments/assets/8803379c-3c14-4a7c-8307-628f926fd4ae" />

fdelbrayelle committed:
fix(parse): arrange code following regression with tika 3.2.2

Force PDFParser for PDF MIME type to avoid empty-output regression
introduced in Tika 3.2.2 with AutoDetectParser + PDFParserConfig.
Fixes kestra-io/kestra#11362.

fdelbrayelle commented:
## 🤖 QA Report — branch `fix/issue-11362-clean` (Kestra 1.3.9 OSS)

**Tested:** `io.kestra.plugin.tika.Parse` · Tika 3.2.2 regression · 4 flows

---

### Summary

| # | Flow | Status | Key Check |
|---|------|--------|-----------|
| 1 | `parse-pdf` (TEXT, store=false) | ✅ SUCCESS | PDF text extracted correctly |
| 2 | `parse-pdf-xhtml-store` (XHTML, store=true) | ✅ SUCCESS | `kestra://` URI returned |
| 3 | `parse-image-metadata` (JPEG + OCR) | ⚠️ REGRESSION | `EmptyParser` — empty content & metadata |
| 4 | `parse-text-file` (plain text, store=false) | ⚠️ REGRESSION | `EmptyParser` — no real content |

---

### Flow 1: parse-pdf ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_text
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    contentType: TEXT
    store: false

  - id: log_extracted_text
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.parse_text.result.content }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 1.5s |
| `parse_text` | SUCCESS | 2.9s |
| `log_extracted_text` | SUCCESS | 0.1s |
| **Total** | **SUCCESS** | **5.0s** |

**Logs synthesis**
Clean execution, no errors or warnings.

**Outputs synthesis**
`parse_text` returned `result.content` with 5 469 characters of real PDF text (e.g. `"BLOG\n\nYour Private App Store for Data Pipelines…"`). `result.metadata` contains standard PDF metadata (`Content-Type: application/pdf`, `pdf:PDFVersion: 1.4`, etc.). No `EmptyParser` in metadata. **PDF regression is fixed. ✅**

---

### Flow 2: parse-pdf-xhtml-store ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf-xhtml-store
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_xhtml
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    store: true

  - id: log_uri
    type: io.kestra.plugin.core.log.Log
    message: "uri={{ outputs.parse_xhtml.uri }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 1.7s |
| `parse_xhtml` | SUCCESS | 2.7s |
| `log_uri` | SUCCESS | 0.1s |
| **Total** | **SUCCESS** | **5.0s** |

**Logs synthesis**
Clean execution, no errors.

**Outputs synthesis**
`parse_xhtml` returned `uri: kestra:///company/team/parse-pdf-xhtml-store/executions/.../318905220877381542.ion`. The Ion file is correctly written to internal storage. **`store: true` works correctly for PDFs. ✅**

---

### Flow 3: parse-image-metadata ⚠️ REGRESSION

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-image-metadata
namespace: company.team

tasks:
  - id: get_image
    type: io.kestra.plugin.core.http.Download
    uri: https://upload.wikimedia.org/wikipedia/commons/a/a7/Camponotus_flavomarginatus_ant.jpg

  - id: tika
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.get_image.uri }}"
    store: false
    contentType: TEXT
    ocrOptions:
      strategy: OCR_AND_TEXT_EXTRACTION

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.tika.result.metadata }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `get_image` | SUCCESS | 0.6s |
| `tika` | SUCCESS | 0.1s |
| `log_metadata` | SUCCESS | 0.03s |
| **Total** | **SUCCESS** | **0.8s** |

**Logs synthesis**
No errors logged — execution completes silently, but `tika` task runs suspiciously fast (70ms), confirming `EmptyParser` is used instead of any real image/OCR parser.

**Outputs synthesis**
```json
{
  "X-TIKA:Parsed-By": "org.apache.tika.parser.EmptyParser",
  "X-TIKA:Parsed-By-Full-Set": "org.apache.tika.parser.EmptyParser",
  "Content-Type": "image/jpeg"
}
```
`result.content` is empty. No EXIF or image metadata extracted. This is identical to the output reported by @Malaydewangan09 in the PR comments. **Image parsing regression is NOT fixed. ❌**

---

### Flow 4: parse-text-file ⚠️ REGRESSION

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-text-file
namespace: company.team

tasks:
  - id: download_text
    type: io.kestra.plugin.core.http.Download
    uri: https://www.gutenberg.org/files/2701/2701-0.txt

  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_text.uri }}"
    extractEmbedded: true
    store: false

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "parsed_by={{ outputs.parse.result.metadata['X-TIKA:Parsed-By'] }} content_length={{ outputs.parse.result.content | length }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_text` | SUCCESS | 2.3s |
| `parse` | SUCCESS | 0.4s |
| `log_metadata` | SUCCESS | 0.1s |
| **Total** | **SUCCESS** | **3.3s** |

**Logs synthesis**
No errors. The downloaded file is ~1.2MB of plain text (Moby Dick from Project Gutenberg), yet the parse step completes in 0.4s — consistent with `EmptyParser` doing nothing.

**Outputs synthesis**
```json
{
  "X-TIKA:Parsed-By": "org.apache.tika.parser.EmptyParser",
  "X-TIKA:Parsed-By-Full-Set": "org.apache.tika.parser.EmptyParser",
  "Content-Type": "text/plain"
}
```
`result.content` is a hollow XHTML wrapper (218 chars) with no actual text. The full 1.2MB plain-text file was silently discarded by `EmptyParser`. **Text file parsing regression is NOT fixed. ❌**

---

### Conclusion

The fix on `fix/issue-11362-clean` correctly resolves the PDF regression by dispatching PDF files explicitly to `PDFParser`. However, the Tika 3.2.2 regression also affects **images** (`image/jpeg`, and likely all image formats) and **plain text files** (`text/plain`). Both still fall through to `EmptyParser`. The same explicit-parser dispatch strategy likely needs to be extended to cover at minimum `image/*` and `text/plain` MIME types.

fdelbrayelle committed:
fix(parse): extend explicit parser dispatch to image/* and text/plain

AutoDetectParser falls back to EmptyParser for image/* and text/plain
MIME types inside the shadow JAR due to ServiceLoader discovery issues
introduced in Tika 3.2.2. Force ImageParser and TXTParser explicitly,
mirroring the existing PDFParser workaround. Also update embedded parser
condition to cover all specific-parser cases, not just PDFParser.

fdelbrayelle committed:
fix(parse): use TesseractOCRParser for images when OCR is requested

ImageParser only extracts metadata and cannot perform OCR, breaking the
image+OCR test case. Hoist ocrStrategy computation before parser selection
and switch to TesseractOCRParser for image/* when OCR strategy is not NO_OCR.

fdelbrayelle committed:
fix(parse): initialize TesseractOCRParser before use and use TikaInputStream

Two root causes fixed:
1. TesseractOCRParser implements Initializable — calling initialize() is
   required before use when not going through TikaConfig; without it,
   hasTesseract() returns false and OCR is silently skipped.
2. Route the input through a file-backed TikaInputStream so Tesseract can
   access the content via a real filesystem path, as it does when invoked
   through AutoDetectParser internally.

fdelbrayelle commented:
## 🤖 QA Report — branch `issue/11362` (Kestra `latest` OSS, 2026-04-10)

**Tested:** `io.kestra.plugin.tika.Parse` · Exact rerun of [previous QA scenarios](#issuecomment-4216011716) · 4 flows

---

### Summary

| # | Flow | Status | Key Check |
|---|------|--------|-----------|
| 1 | `parse-pdf` (TEXT, store=false) | ✅ SUCCESS | PDF text extracted correctly |
| 2 | `parse-pdf-xhtml-store` (XHTML, store=true) | ✅ SUCCESS | `kestra://` URI returned |
| 3 | `parse-image-metadata` (JPEG + OCR) | ⚠️ REGRESSION | Empty content & minimal metadata — Tesseract not in container |
| 4 | `parse-text-file` (plain text, store=false) | ✅ FIXED | 1,219,400 chars extracted — **regression resolved** |

---

### Flow 1: parse-pdf ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_text
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    contentType: TEXT
    store: false

  - id: log_extracted_text
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.parse_text.result.content }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 1.2s |
| `parse_text` | SUCCESS | 0.5s |
| `log_extracted_text` | SUCCESS | 0.03s |
| **Total** | **SUCCESS** | **2.3s** |

**Logs synthesis**
Clean execution, no errors or warnings. PDF downloaded (3.2MB), extracted text logged successfully.

**Outputs synthesis**
`parse_text` returned `result.content` with real PDF text starting with `"BLOG\n\nYour Private App Store for Data Pipelines…"`. `result.metadata` contains `Content-Type: application/pdf`. No `EmptyParser` in metadata. **PDF parsing ✅ confirmed working.**

---

### Flow 2: parse-pdf-xhtml-store ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf-xhtml-store
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_xhtml
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    store: true

  - id: log_uri
    type: io.kestra.plugin.core.log.Log
    message: "uri={{ outputs.parse_xhtml.uri }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 1.2s |
| `parse_xhtml` | SUCCESS | 0.5s |
| `log_uri` | SUCCESS | 0.06s |
| **Total** | **SUCCESS** | **2.3s** |

**Logs synthesis**
Clean execution, no errors.

**Outputs synthesis**
`parse_xhtml` returned a valid `uri: kestra:///company/team/parse-pdf-xhtml-store/executions/.../....ion` pointing to internal storage. **`store: true` ✅ confirmed working for PDFs.**

---

### Flow 3: parse-image-metadata ⚠️ REGRESSION (unchanged)

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-image-metadata
namespace: company.team

tasks:
  - id: get_image
    type: io.kestra.plugin.core.http.Download
    uri: https://upload.wikimedia.org/wikipedia/commons/a/a7/Camponotus_flavomarginatus_ant.jpg

  - id: tika
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.get_image.uri }}"
    store: false
    contentType: TEXT
    ocrOptions:
      strategy: OCR_AND_TEXT_EXTRACTION

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.tika.result.metadata }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `get_image` | SUCCESS | 1.0s |
| `tika` | SUCCESS | 0.24s |
| `log_metadata` | SUCCESS | 0.04s |
| **Total** | **SUCCESS** | **1.8s** |

**Logs synthesis**
No errors logged. The `tika` task completes in 240ms (suspiciously fast for OCR+image parsing), and the log step outputs only `{"Content-Type":"image/jpeg"}` — consistent with a no-op parse.

**Outputs synthesis**
```json
{
  "Content-Type": "image/jpeg"
}
```
`result.content` is empty. No EXIF metadata, no `X-TIKA:Parsed-By`, no image dimensions — only `Content-Type`. This is effectively the same EmptyParser outcome as the previous QA run.

> **Root cause observed:** `tesseract` is not installed in the `kestra/kestra:latest` Docker image (`exec: "tesseract": executable file not found in $PATH`). The `TesseractOCRParser` initialization likely fails silently and falls back to no-op parsing. Even basic JPEG metadata (EXIF: width, height, etc.) is not extracted, suggesting the image parser dispatch is still not reaching a functional parser. **Image parsing regression is NOT fixed. ❌**

---

### Flow 4: parse-text-file ✅ FIXED

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-text-file
namespace: company.team

tasks:
  - id: download_text
    type: io.kestra.plugin.core.http.Download
    uri: https://www.gutenberg.org/files/2701/2701-0.txt

  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_text.uri }}"
    extractEmbedded: true
    store: false

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "metadata={{ outputs.parse.result.metadata }} content_length={{ outputs.parse.result.content | length }}"
```
</details>

> **Note:** The original log template from the previous QA (`outputs.parse.result.metadata['X-TIKA:Parsed-By']`) caused the flow to FAIL because `X-TIKA:Parsed-By` is no longer present in the metadata (Tika 3.x changed the metadata key structure). The template was updated to log all metadata safely.

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_text` | SUCCESS | 1.8s |
| `parse` | SUCCESS | 0.06s |
| `log_metadata` | SUCCESS | 0.24s |
| **Total** | **SUCCESS** | **2.7s** |

**Logs synthesis**
Clean execution, no errors. Log step outputs:
```
metadata={"Content-Encoding":"UTF-8","X-TIKA:detectedEncoding":"UTF-8","X-TIKA:encodingDetector":"UniversalEncodingDetector","Content-Type":"text/plain; charset=UTF-8"} content_length=1219400
```
A real text parser (`UniversalEncodingDetector`) is now being used. No `EmptyParser` in metadata.

**Outputs synthesis**
`parse.result.content` contains **1,219,400 characters** of real text from Moby Dick (previously 218 chars of hollow XHTML with EmptyParser). `parse.result.metadata` now contains encoding detection fields. **Text file parsing regression is FIXED ✅**

---

### Summary of changes since previous QA

| Regression | Previous | Now |
|------------|----------|-----|
| PDF parsing | ✅ FIXED (was already fixed) | ✅ Still fixed |
| Text file (`text/plain`) parsing | ❌ EmptyParser (0 chars) | ✅ FIXED (1,219,400 chars) |
| Image (`image/jpeg`) parsing | ❌ EmptyParser | ❌ Still EmptyParser |

The `text/plain` regression introduced by Tika 3.2.2 is resolved. The `image/*` regression remains open — the `TesseractOCRParser`-based fix does not produce results in the standard Kestra Docker image (which lacks Tesseract), and the plugin also does not fall back to basic image metadata extraction (EXIF) when OCR is unavailable.

fdelbrayelle committed:
fix(parse): fall back to ImageParser when Tesseract is unavailable for images

When OCR is requested for image/* files but the tesseract binary is not
installed, TesseractOCRParser produced completely empty output (no content,
no metadata). Guard the TesseractOCRParser dispatch behind hasTesseract()
and fall back to ImageParser so that at minimum EXIF metadata is extracted.

fdelbrayelle commented:
## 🤖 QA Report — branch `issue/11362` (Kestra `latest` OSS, 2026-04-10)

**Tested:** `io.kestra.plugin.tika.Parse` · Exact rerun of [previous QA scenarios](#issuecomment-4221589370) · 4 flows

---

### Summary

| # | Flow | Status | Key Check |
|---|------|--------|-----------|
| 1 | `parse-pdf` (TEXT, store=false) | ✅ SUCCESS | PDF text extracted correctly |
| 2 | `parse-pdf-xhtml-store` (XHTML, store=true) | ✅ SUCCESS | `kestra://` URI returned |
| 3 | `parse-image-metadata` (JPEG + OCR) | ✅ **FIXED** | Rich JPEG metadata returned — `ImageParser` fallback working |
| 4 | `parse-text-file` (plain text, store=false) | ✅ SUCCESS | 1,219,400 chars extracted |

---

### Flow 1: parse-pdf ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_text
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    contentType: TEXT
    store: false

  - id: log_extracted_text
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.parse_text.result.content }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 0.73s |
| `parse_text` | SUCCESS | 0.29s |
| `log_extracted_text` | SUCCESS | 0.03s |
| **Total** | **SUCCESS** | **1.1s** |

**Logs synthesis**
Clean execution, no errors or warnings. PDF downloaded (3.2MB), extracted text logged successfully.

**Outputs synthesis**
`parse_text` returned `result.content` with real PDF text starting with `"BLOG\n\nYour Private App Store for Data Pipelines…"`. Metadata includes `Content-Type: application/pdf`, `xmpTPg:NPages: 2`, PDF version 1.4, producer Skia/PDF. No `EmptyParser`. **PDF parsing ✅ confirmed working.**

---

### Flow 2: parse-pdf-xhtml-store ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf-xhtml-store
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_xhtml
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    store: true

  - id: log_uri
    type: io.kestra.plugin.core.log.Log
    message: "uri={{ outputs.parse_xhtml.uri }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 0.73s |
| `parse_xhtml` | SUCCESS | 0.32s |
| `log_uri` | SUCCESS | 0.03s |
| **Total** | **SUCCESS** | **1.1s** |

**Logs synthesis**
Clean execution, no errors.

**Outputs synthesis**
`parse_xhtml` returned a valid `uri: kestra:///company/team/parse-pdf-xhtml-store/executions/.../....ion` pointing to internal storage. **`store: true` ✅ confirmed working for PDFs.**

---

### Flow 3: parse-image-metadata ✅ FIXED

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-image-metadata
namespace: company.team

tasks:
  - id: get_image
    type: io.kestra.plugin.core.http.Download
    uri: https://upload.wikimedia.org/wikipedia/commons/a/a7/Camponotus_flavomarginatus_ant.jpg

  - id: tika
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.get_image.uri }}"
    store: false
    contentType: TEXT
    ocrOptions:
      strategy: OCR_AND_TEXT_EXTRACTION

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.tika.result.metadata }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `get_image` | SUCCESS | 0.48s |
| `tika` | SUCCESS | 0.28s |
| `log_metadata` | SUCCESS | 0.04s |
| **Total** | **SUCCESS** | **0.8s** |

**Logs synthesis**
No errors logged. `tika` task now takes 280ms (compared to 240ms previously), consistent with actually invoking a real image parser instead of falling through to a no-op.

**Outputs synthesis**
```json
{
  "Content-Type": "image/jpeg",
  "width": "1800",
  "height": "1200",
  "tiff:ImageWidth": "1800",
  "tiff:ImageLength": "1200",
  "Chroma ColorSpaceType": "YCbCr",
  "Chroma NumChannels": "3",
  "Compression CompressionTypeName": "JPEG",
  "Compression Lossless": "false",
  "Compression NumProgressiveScans": "1",
  "Dimension HorizontalPixelSize": "0.25148514",
  "Dimension VerticalPixelSize": "0.25148514",
  "Dimension ImageOrientation": "normal",
  "Dimension PixelAspectRatio": "1.0",
  "Document ImageCreationTime": "year=2009, month=1, day=27, hour=15, minute=29, second=8",
  "JPEGvariety app0JFIF": "majorVersion=1, minorVersion=2, resUnits=1, Xdensity=101, Ydensity=101, ...",
  "imagereader:NumImages": "1",
  ...
}
```
`result.content` remains empty (Tesseract not installed in `kestra/kestra:latest`), but **metadata is now fully populated with 20+ fields** including image dimensions (1800×1200), color space, DPI, compression details, and image creation date.

> **Previous run**: only `{"Content-Type":"image/jpeg"}` — no-op parse, EmptyParser behavior.
> **This run**: rich metadata via `ImageParser` fallback. **Image metadata regression is FIXED ✅** (commit `edfcfb5`).

---

### Flow 4: parse-text-file ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-text-file
namespace: company.team

tasks:
  - id: download_text
    type: io.kestra.plugin.core.http.Download
    uri: https://www.gutenberg.org/files/2701/2701-0.txt

  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_text.uri }}"
    extractEmbedded: true
    store: false

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "metadata={{ outputs.parse.result.metadata }} content_length={{ outputs.parse.result.content | length }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_text` | SUCCESS | 1.58s |
| `parse` | SUCCESS | 0.09s |
| `log_metadata` | SUCCESS | 0.25s |
| **Total** | **SUCCESS** | **1.9s** |

**Logs synthesis**
Clean execution, no errors. Log step outputs:
```
metadata={"Content-Encoding":"UTF-8","X-TIKA:detectedEncoding":"UTF-8","X-TIKA:encodingDetector":"UniversalEncodingDetector","Content-Type":"text/plain; charset=UTF-8"} content_length=1219400
```

**Outputs synthesis**
`parse.result.content` contains **1,219,400 characters** of real text from Moby Dick. `UniversalEncodingDetector` in metadata confirms a real text parser is being used. **Text file parsing ✅ confirmed stable.**

---

### Verdict

All 4 scenarios pass. The image metadata regression introduced by Tika 3.2.2 is **fully resolved** by the `ImageParser` fallback added in commit `edfcfb5`: when Tesseract is unavailable, the parser now falls back to `ImageParser` and returns complete JPEG/image metadata instead of a bare `Content-Type`. Text OCR content remains empty (as expected without Tesseract), but metadata extraction works correctly.

fdelbrayelle committed:
chore: update dead link to picture

fdelbrayelle committed:
fix(parse): emit image metadata as content when ImageParser produces no text

ImageParser.parse() only populates Metadata (EXIF, dimensions, colour
space…) and writes nothing to the ContentHandler, leaving result.content
empty.  Add a useImageParserFallback flag that is set whenever ImageParser
is the effective parser (NO_OCR path and Tesseract-unavailable fallback),
then build a sorted key: value text from the Metadata entries as fallback
content.  When Tesseract is available the OCR-produced handler content is
used as-is.  Add a test case asserting tiff:ImageWidth appears in content
for a plain-image parse.

fdelbrayelle commented:
## 🤖 QA Report — branch `issue/11362` (Kestra `latest` OSS, 2026-04-10)

**Tested:** `io.kestra.plugin.tika.Parse` · Exact rerun of [previous QA scenarios](#issuecomment-4221769887) · 4 flows

---

### Summary

| # | Flow | Status | Key Check |
|---|------|--------|-----------|
| 1 | `parse-pdf` (TEXT, store=false) | ✅ SUCCESS | PDF text extracted correctly |
| 2 | `parse-pdf-xhtml-store` (XHTML, store=true) | ✅ SUCCESS | `kestra://` URI returned |
| 3 | `parse-image-metadata` (JPEG + OCR) | ✅ SUCCESS | Rich JPEG metadata returned — `ImageParser` fallback confirmed |
| 4 | `parse-text-file` (plain text, store=false) | ✅ SUCCESS | 1,219,400 chars extracted |

---

### Flow 1: parse-pdf ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_text
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    contentType: TEXT
    store: false

  - id: log_extracted_text
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.parse_text.result.content }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 0.99s |
| `parse_text` | SUCCESS | 2.15s |
| `log_extracted_text` | SUCCESS | 0.11s |
| **Total** | **SUCCESS** | **3.2s** |

**Logs synthesis**
Clean execution, no warnings or errors.

**Outputs synthesis**
`parse_text` returned `result.content` with real PDF text starting with `"\nBLOG\n\nYour Private App Store for Data Pipelines…"`. Metadata includes `pdf:PDFVersion: 1.4`, `Content-Type: application/pdf`, `xmp:CreatorTool: Chrome/116.0.0.0`. **PDF text extraction ✅ confirmed working.**

---

### Flow 2: parse-pdf-xhtml-store ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-pdf-xhtml-store
namespace: company.team

tasks:
  - id: download_pdf
    type: io.kestra.plugin.core.http.Download
    uri: https://huggingface.co/datasets/kestra/datasets/resolve/main/pdf/app_store.pdf

  - id: parse_xhtml
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_pdf.uri }}"
    store: true

  - id: log_uri
    type: io.kestra.plugin.core.log.Log
    message: "uri={{ outputs.parse_xhtml.uri }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 1.04s |
| `parse_xhtml` | SUCCESS | 2.15s |
| `log_uri` | SUCCESS | 0.14s |
| **Total** | **SUCCESS** | **3.3s** |

**Logs synthesis**
Clean execution, no warnings or errors.

**Outputs synthesis**
`parse_xhtml` returned a valid `uri: kestra:///company/team/parse-pdf-xhtml-store/executions/…` pointing to internal storage. **`store: true` ✅ confirmed working for PDFs.**

---

### Flow 3: parse-image-metadata ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-image-metadata
namespace: company.team

tasks:
  - id: get_image
    type: io.kestra.plugin.core.http.Download
    uri: https://upload.wikimedia.org/wikipedia/commons/a/a7/Camponotus_flavomarginatus_ant.jpg

  - id: tika
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.get_image.uri }}"
    store: false
    contentType: TEXT
    ocrOptions:
      strategy: OCR_AND_TEXT_EXTRACTION

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "{{ outputs.tika.result.metadata }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `get_image` | SUCCESS | 0.61s |
| `tika` | SUCCESS | 1.14s |
| `log_metadata` | SUCCESS | 0.20s |
| **Total** | **SUCCESS** | **1.95s** |

**Logs synthesis**
No errors or warnings. `tika` task completed in 1.14s.

**Outputs synthesis**
`tika` returned **26 metadata fields** including:
```json
{
  "Content-Type": "image/jpeg",
  "height": "1200",
  "tiff:ImageLength": "1200",
  "markerSequence sof": "process=0, samplePrecision=8, numLines=1200, samplesPerLine=1800, numFrameComponents=3",
  "Chroma ColorSpaceType": "YCbCr",
  "Chroma NumChannels": "3",
  "Compression CompressionTypeName": "JPEG",
  "JPEGvariety app0JFIF": "majorVersion=1, minorVersion=2, resUnits=1, Xdensity=101, Ydensity=101, ...",
  "Dimension VerticalPixelSize": "0.25148514",
  "markerSequence dri": "225",
  ...
}
```
`result.content` contains text representation of JPEG metadata fields (Tesseract not installed in `kestra/kestra:latest`, so no OCR text). **Image metadata (`ImageParser` fallback) ✅ confirmed stable.**

---

### Flow 4: parse-text-file ✅ SUCCESS

<details>
<summary>Flow YAML</summary>

```yaml
id: parse-text-file
namespace: company.team

tasks:
  - id: download_text
    type: io.kestra.plugin.core.http.Download
    uri: https://www.gutenberg.org/files/2701/2701-0.txt

  - id: parse
    type: io.kestra.plugin.tika.Parse
    from: "{{ outputs.download_text.uri }}"
    extractEmbedded: true
    store: false

  - id: log_metadata
    type: io.kestra.plugin.core.log.Log
    message: "metadata={{ outputs.parse.result.metadata }} content_length={{ outputs.parse.result.content | length }}"
```
</details>

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_text` | SUCCESS | 3.23s |
| `parse` | SUCCESS | 0.47s |
| `log_metadata` | SUCCESS | 0.69s |
| **Total** | **SUCCESS** | **4.4s** |

**Logs synthesis**
Clean execution, no warnings or errors. Log step outputs metadata and content length.

**Outputs synthesis**
`parse.result.content` contains **1,219,400 characters** of text (Moby Dick). Metadata: `Content-Encoding: UTF-8`, `X-TIKA:detectedEncoding: UTF-8`, `X-TIKA:encodingDetector: UniversalEncodingDetector`, `Content-Type: text/plain; charset=UTF-8`. **Text file parsing ✅ confirmed stable.**

---

### Verdict

All 4 scenarios pass. The `ImageParser` fallback introduced in commit `edfcfb5` continues to work correctly: JPEG metadata returns 26 fields (vs bare `Content-Type` before the fix). All other scenarios remain stable.

fdelbrayelle commented:
## 🤖 QA Report — branch `issue/11362` (Kestra `latest` OSS, 2026-04-10) — post-simplification

**Tested:** `io.kestra.plugin.tika.Parse` · Same 4 scenarios · after removing `TesseractOCRParser` detection for images

**Change**: the `image/*` parser-selection branch was simplified to always use `ImageParser` unconditionally, removing the `TesseractOCRParser` instantiation / `hasTesseract()` / fallback chain. Kestra's standard image (`kestra/kestra:latest`) does not ship Tesseract, so that path was dead code. Image metadata extraction does not require Tesseract.

---

### Summary

| # | Flow | Status | Key Check |
|---|------|--------|-----------|
| 1 | `parse-pdf` (TEXT, store=false) | ✅ SUCCESS | PDF text extracted correctly |
| 2 | `parse-pdf-xhtml-store` (XHTML, store=true) | ✅ SUCCESS | `kestra://` URI returned |
| 3 | `parse-image-metadata` (JPEG, OCR_AND_TEXT_EXTRACTION) | ✅ SUCCESS | 26 metadata fields, no Tesseract needed |
| 4 | `parse-text-file` (plain text, store=false) | ✅ SUCCESS | 1,219,400 chars extracted |

---

### Flow 1: parse-pdf ✅ SUCCESS

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 0.82s |
| `parse_text` | SUCCESS | 0.53s |
| `log` | SUCCESS | 0.04s |

`result.content` starts: `"\nBLOG\n\nYour Private App Store for Data Pipelines…"` ✅

---

### Flow 2: parse-pdf-xhtml-store ✅ SUCCESS

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_pdf` | SUCCESS | 0.75s |
| `parse_xhtml` | SUCCESS | 0.53s |
| `log` | SUCCESS | 0.03s |

`uri: kestra:///…` returned ✅

---

### Flow 3: parse-image-metadata ✅ SUCCESS

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `get_image` | SUCCESS | 0.48s |
| `tika` | SUCCESS | 0.39s |
| `log` | SUCCESS | 0.05s |

26 metadata fields: `Content-Type: image/jpeg`, `width: 1800`, `height: 1200`, `Chroma ColorSpaceType: YCbCr`, `Compression CompressionTypeName: JPEG`, etc. `result.content` is a plain-text representation of those fields (metadata-as-content fallback). **No Tesseract required ✅**

> Note: `ocrOptions.strategy: OCR_AND_TEXT_EXTRACTION` was set in the flow. With the simplified code, this no longer attempts to invoke Tesseract — it is silently ignored for images and `ImageParser` is used directly. If true OCR text extraction from images is needed in the future, Tesseract support can be added as a separate feature in a follow-up PR, with clear documentation that it requires Tesseract to be installed on the host.

---

### Flow 4: parse-text-file ✅ SUCCESS

**Gantt**
| Task | Status | Duration |
|------|--------|----------|
| `download_text` | SUCCESS | 5.45s |
| `parse` | SUCCESS | 0.09s |
| `log` | SUCCESS | 0.34s |

`content_length: 1,219,400` ✅

---

### Verdict

All 4 scenarios pass after removing the Tesseract detection logic from the `image/*` path. Image metadata extraction is clean, dependency-free, and produces identical results to the previous run.

fdelbrayelle committed:
refactor(parse): simplify image OCR branch — use TesseractOCRParser when available, ImageParser otherwise

When OCR_AND_TEXT_EXTRACTION is requested for image/* files:
- use TesseractOCRParser (initialize + hasTesseract check) if Tesseract is on PATH
- fall back to ImageParser (metadata-as-content) if Tesseract is absent

NO_OCR always uses ImageParser directly. Tesseract is never a hard dependency.

fdelbrayelle committed:
docs(parse): document Tesseract as prerequisite for image OCR

Clarify in the class @Schema, the OCR example title, and the ocrOptions
field description that Tesseract must be installed on the Kestra host
for OCR_AND_TEXT_EXTRACTION to produce text from images. Without it,
the plugin falls back to ImageParser (metadata only).
